Extend basic block without aggressive memory copy

qwe661234 opened the pull request:
Due to the significant overhead associated with using aggressive memory copy, we opted to use a pointer to the IR of the basic block we intend to merge, rather than copying the IR to the basic block we wish to extend. 

The performance results obtained from running CoreMark using different implementation strategies for EBB are presented below. As we can see, aggressive memory copy substantially degrades performance.

* Model:  Core i7-8700

|Test| aggressive memcpy|Compiler | Iterations / Sec |Speedup|
|--------| -------- |-------- | -------- |-------- |
| BB  | |clang-15 |  971.951     |       |
| BB  | |gcc-12   |  963.336     |       |
| EBB  |O|clang-15 |  1013.070|+4.2% |
| EBB   |O |gcc-12   |  1020.391    | +6%|
| EBB | X|clang-15 |  1160.894    |+19.4% |
| EBB|X |gcc-12   |  1167.938    | +21.2%|

jserv commented on src/decode.h:
```diff
@@ -259,6 +259,13 @@ typedef struct rv_insn {
      */
     bool tailcall;
     bool (*impl)(riscv_t *, const struct rv_insn *);
+    /* The pointer branch take points to the first ir of the first basic block
```
Improve the writing. ChatGPT is capable of proofreading.

jserv commented on src/decode.h:
```diff
@@ -259,6 +259,13 @@ typedef struct rv_insn {
+     * pointer allows us to extend the basic block without using memcpy to copy
+     * ir array.
+     */
+    struct rv_insn *branch_take, *branch_not_take;
```
Should be `branch_taken` and `branch_untaken`

jserv reviewed (commented):
Improve git commit message. Always mention "IR" in capitalized form. You shall show explicit numbers about memory consumption and potential benefits.

jserv commented on src/emulate.c:
```diff
@@ -1340,6 +1422,27 @@ static void block_translate(riscv_t *rv, block_t *block)
+             not_taken_pc = block->pc_end;
+
+    block_map_t *map = &rv->block_map;
+    if (!last_ir->branch_take) {
```
Attempt to simplify the logic.

jserv commented on src/decode.h:
```diff
@@ -259,6 +259,12 @@ typedef struct rv_insn {
      */
     bool tailcall;
     bool (*impl)(riscv_t *, const struct rv_insn *);
+    /* To extend the basic block without frequently using memory copy to copy
```
Avoid writing annoying words such as "using memory copy to copy."

jserv commented on src/emulate.c:
```diff
@@ -341,90 +338,139 @@ RVOP(jalr, {
     const uint32_t pc = rv->PC;
-    if (rv->X[ir->rs1] == rv->X[ir->rs2])
+    if (rv->X[ir->rs1] == rv->X[ir->rs2]) {
+        if (ir->branch_untaken) {
```
Rewrite as following:
```c
    if (!ir->branch_untaken)
        goto nextop;
    rv->PC += ir->insn_len;
    return ir->branch_untaken->impl(rv, ir->branch_untaken);
```

jserv commented on src/emulate.c:
```diff
@@ -341,90 +338,139 @@ RVOP(jalr, {
     const uint32_t pc = rv->PC;
-    if ((int32_t) rv->X[ir->rs1] >= (int32_t) rv->X[ir->rs2])
+    if ((int32_t) rv->X[ir->rs1] >= (int32_t) rv->X[ir->rs2]) {
+        if (ir->branch_untaken) {
```
Ditto. Shorten the code.

jserv commented on src/decode.h:
```diff
@@ -259,6 +259,14 @@ typedef struct rv_insn {
     bool tailcall;
     bool (*impl)(riscv_t *, const struct rv_insn *);
+
+    /* To avoid the significant overhead that comes with using aggressive memory
```
Rephrase:
> We employ two pointers, `branch_taken` and `branch_untaken`, to avoid the significant overhead resulting from aggressive memory copy.

jserv commented on src/emulate.c:
```diff
@@ -1340,6 +1414,23 @@ static void block_translate(riscv_t *rv, block_t *block)
     block->ir[block->n_insn - 1].tailcall = true;
 }
 
+static void extend_block(riscv_t *rv, block_t *block)
```
Add some comments and/or TODO.

jserv replied:
You shall describe the considerations and the idea behind the approach you are taking.

qwe661234 committed:
Extend basic block without aggressive memory copy

Due to the significant overhead associated with using aggressive memory
copy, we opted to use a pointer to the begin of IR array of the basic block
we intend to merge, rather than copying whole IR array to the basic block
we wish to extend.

The performance results obtained from running CoreMark using different
implementation strategies for EBB are presented below. As we can see,
aggressive memory copy substantially degrades performance.

|Test| aggressive memcpy|Compiler | Iterations / Sec |Speedup|
|----+------------------+---------+------------------+-------|
| BB |                  |clang-15 |  971.951         |       |
|----+------------------+---------+------------------+-------|
| BB |                  | gcc-12  |  963.336         |       |
|----+------------------+---------+------------------+-------|
| EBB|       O          |clang-15 |  1013.070        | +4.2% |
|----+------------------+---------+------------------+-------|
| EBB|       O          | gcc-12  |  1020.391        | +6%   |
|----+------------------+---------+------------------+-------|
| EBB|       X          |clang-15 |  1160.894        | +19.4%|
|----+------------------+---------+------------------+-------|
| EBB|       X          | gcc-12  | 1167.938         | +21.2%|
